openhands/server/conversation_manager/docker_nested_conversation_manager.py

```diff
@@ -281,7 +281,23 @@ async def disconnect_from_session(self, connection_id: str):
         raise ValueError('unsupported_operation')
 
     async def close_session(self, sid: str):
-        stop_all_containers(f'openhands-runtime-{sid}')
+        # First try to graceful stop server.
+        try:
+            container = self.docker_client.containers.get(f'openhands-runtime-{sid}')
+        except docker.errors.NotFound as e:
+            return
+        try:
+            nested_url = self.get_nested_url_for_container(container)
+            async with httpx.AsyncClient(
+                headers={
+                    'X-Session-API-Key': self._get_session_api_key_for_conversation(sid)
+                }
+            ) as client:
+                response = await client.post(f'{nested_url}/api/conversations/{sid}/stop')
+                response.raise_for_status()
+        except Exception:
+            logger.error("error_stopping_container")
```

Comment:
If we do logger.exception here it will attach stacktrace (same as exc_info=True), but that can wait for later PR especially since this code is turned off still.